File: 1.2/similarity.swift
import Foundation

let inputText = try! String(contentsOfFile: "input.txt", encoding: String.Encoding.utf8)
let lines = inputText.components(separatedBy: CharacterSet.whitespacesAndNewlines)

var l: [Int] = []
var r: [Int] = []

for i in 0..<lines.count {
  if (lines[i]) == "" {
    continue
  }
  if i % 2 == 0 {
    l.append(Int(lines[i])!)
  } else {
    r.append(Int(lines[i])!)
  }
}

l.sort()
r.sort()

var total = 0, ridx = 0;

for lidx in 0..<l.count {
  var rcount = 0;

  while r[ridx] < l[lidx] {
    ridx += 1;
  }

  while r[ridx] == l[lidx] {
    rcount += 1;
    ridx += 1;
  }

  total += l[lidx] * rcount
}

print(total)

File: 1.2/similarity.rs
use std::fs::read_to_string;

fn parse_lines() -> (Vec<i32>, Vec<i32>){
    let mut l = Vec::new();
    let mut r = Vec::new();

    for line in read_to_string("input.txt").unwrap().lines() {
        let Some((left, right)) = line.split_once("   ") else { panic!("Bad input line!") };
        l.push(left.parse::<i32>().unwrap());
        r.push(right.parse::<i32>().unwrap());
    }

    l.sort();
    r.sort();

    return (l, r);
}

fn main() {
    let lists = parse_lines();

    let mut total = 0;
    for lidx in 0..lists.0.len() {
        let mut rcount = 0;
        let mut ridx = 0;

        while lists.1[ridx] < lists.0[lidx] {
            ridx += 1;
        }

        while lists.1[ridx] == lists.0[lidx] {
            rcount += 1;
            ridx += 1;
        }

        total += lists.0[lidx] * rcount;
    }

    println!("{:?}", total);
}